### DISCUSSION TITLE
Error while trying to use a plugin outside the reach folder

### DISCUSSION THREAD

**Discussion by vicltz** [ros-industrial/reach#61] [Q&A] [answered]
Hi,

I would like to use a target_pose_generator plugin with reach but outside the reach folder.

---------------------------------------------------------
**What works**

I wrote a `target_pose_generator` plugin. Firstly, I was writing it inside the reach folder on my machine.
I was using the following (working) setup.

I added these lines in the `plugins.cpp` file:

```
#include <reach/plugins/test_plugin.h>
EXPORT_TARGET_POSE_GENERATOR_PLUGIN(reach::TestPluginFactory, TestPlugin)
```

And this in the reach CMakelist:
```

add_library(
  ${PROJECT_NAME} SHARED
  # other plugins ...
  # My Plugin !
  src/plugins/test_plugin.cpp

```

And then I called it in the `reach_study.yaml` file. All of that worked perfectly.

---------------------------------------

**What I would like to do**

The thing I would like to do is to bundle this plugin in another package (that would be in the same catkin workspace than reach and reach_ros) .

**What I've tried** 

My plugin code is in `my_package/src/test_plugin.cpp` and `my_package/include/test_plugin.h`. I also created a `src/plugin.cpp` file which is the following:

```
#include <reach/plugin_utils.h>
#include "test_plugin.h"


EXPORT_TARGET_POSE_GENERATOR_PLUGIN(reach::TestPluginFactory,TestPlugin)
```

To do that I got inspired by the reach_ros CMakeList. Here is my CMakeList:
```
cmake_minimum_required(VERSION 3.10.0)

# Extract package name and version from package.xml
find_package(ros_industrial_cmake_boilerplate REQUIRED)
extract_package_metadata(pkg)
project(${pkg_extracted_name} VERSION ${pkg_extracted_version} LANGUAGES CXX)

find_package(reach)
find_package(catkin REQUIRED COMPONENTS reach roscpp)

catkin_package(  
  #LIBRARIES ${PROJECT_NAME}
  CATKIN_DEPENDS roscpp boost eigen
)


include_directories(include)

include_directories(
  SYSTEM
  ${catkin_INCLUDE_DIRS}
  ${reach_INCLUDE_DIRS}
)

add_library(
  ${PROJECT_NAME}
  src/test_plugin.cpp
)
target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES} reach::reach)


add_library(${PROJECT_NAME}_plugins SHARED src/plugins.cpp)
target_link_libraries(${PROJECT_NAME}_plugins ${PROJECT_NAME})
target_cxx_version(${PROJECT_NAME}_plugins PUBLIC VERSION 14)

install(
  TARGETS ${PROJECT_NAME}  ${PROJECT_NAME}_plugins
  LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
  RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION})

foreach(dir config model results)
install(DIRECTORY ${dir}/ DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/${dir})
endforeach()
```

**What doesn't work**

With this setup, the workspace compiles successfully. However, when I try to use the plugin, I get the following error while trying to run the reach study node:

```
Failed to create plugin instance 'TestPlugin' of type 'reach::TargetPoseGeneratorFactory'
Search Paths (including system folders)
Search Libraries:
    - libreach_plugins.so
    - libreach_ros_plugins.so
Available plugins of type 'reach::TargetPoseGeneratorFactory':
    - PointCloudTargetPoseGenerator
    - TransformedPointCloudTargetPoseGenerator

```
My workspace is sourced after compilation so this is not the problem. 

**Do you have any idea about what could be the issue ? Am I doing anything wrong ?** 

Thank you for your help.

**Comment by marip8** [comment, marked as answer]
Did you set the `REACH_PLUGINS` environment variable per [this section of the README](https://github.com/ros-industrial/reach#plugins)? This is how the plugin loader knows what libraries to look in for plugins. You don't have to set this for plugins defined in the `reach` library because I've defined [compile definitions](https://github.com/ros-industrial/reach/blob/ec6e21b0bb37b0a86e5297576221f71d54a23588/CMakeLists.txt#L75-L76) to define the library and then [add it manually to the plugin loader](https://github.com/ros-industrial/reach/blob/ec6e21b0bb37b0a86e5297576221f71d54a23588/src/reach_study.cpp#L259-L262). For the `reach_ros` demo, I [set this environment variable in the start launch file](https://github.com/ros-industrial/reach_ros/blob/246b2b003b56484426bf3ad3c74e2a697e088d40/launch/start.launch#L11)

**Reply by vicltz** [reply]
Ok, I have found the problem. Yes, I defined the `REACH_PLUGINS` in the terminal before launching the `reach_ros` node. It looks like the reach_ros launch file was overwriting the environment variable. I edited the file and now it seems to work.

However it's probably not the best solution since I had to edit the reach_ros specific file. I could create a new launch file in my package. Let me know if you know any better solution.

Thank you for your help.

**Reply by marip8** [reply]
I'm not opposed to changing the launch file to use an argument to populate the `REACH_PLUGINS` environment variable if you think that makes it more flexible.

**Reply by vicltz** [reply]
It can be easier , for sure ! It can be great if it appends the plugin to the REACH_PLUGINS environment variable (So the user doesn't have to add `reach_ros_plugins` but only his plugin) . Because right now I think it can be confusing if a user doesn't know what's happening in the launch file the variable will be basically overwrited when launching `reach_ros.`

**Reply by gavanderhoorn** [reply]
Would appending to the variable, instead of overwriting, be an option? That way the launch file provides a default, which the user could append to.

---

Edit: ah, I just now see @vicltz already suggests this.

**Reply by vicltz** [reply]
Also, do you think that being able to append in this param `<rosparam param="source_list">[reach_joints]</rosparam>` in `reach_ros ` `setup.launch ` could be an option ? I am actually writing a node that moves the robot from a marker to another and I had to add my node there. 

It could be nice as a parameter.

**Comment by vicltz** [comment]
As discussed earlier, the error of the plugin not found by reach is now solved. However, I still have an issue related to the plugin being outside the reach folder. That's why I'll continue in this discussion.

My plugin is basically inspired from the `point_cloud_target_pose_generator`. It inherits from the `target_pose_generator` interface. 

My factory has a create method, working exactly on the same structure than `point_cloud_target_pose_generator`

```
TargetPoseGenerator::ConstPtr TestPluginFactory::create(const YAML::Node& config) const
{
  return std::make_shared<TestPlugin>(config);
}
```
However, this method only work when the plugin is inside the `reach` folder.

When I try to use it outside of the `reach` folder, It raises a _segfault_. I used _gdb_ to backtrace the error and it comes from `reach_study.cpp` . In these [lines](https://github.com/ros-industrial/reach/blob/ec6e21b0bb37b0a86e5297576221f71d54a23588/src/reach_study.cpp#L295C1-L299C4) . The` ConstPtr` is a null pointer, and this is why the segfault happen.

The only way I found to fix the error is to put this line
`auto factory = loader.createInstance<TargetPoseGeneratorFactory>(get<std::string>(pose_gen_config, "name"));` outside of the method .

I have no clue why this happen but I am pretty sure that there is no difference between the code outside the folder and inside the folder.

Do you think something might be wrong with the reach code ? Or you are sure the error comes from my code ?

Thanks a lot for your help.

**Reply by marip8** [reply]
https://github.com/ros-industrial/reach/blob/ec6e21b0bb37b0a86e5297576221f71d54a23588/src/reach_study.cpp#L295-L299

Which of these lines specifically is causing the segfault? If the segfault actually occurs after these lines, it could be the case that the pointer created by your factory goes out of scope when the code leaves the scope defined by the curly braces

**Reply by vicltz** [reply]
For some reason, I can't backtrace to the exact line, I can only backtrace up to this line.

`
reach::runReachStudy(YAML::Node const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, boost::filesystem::path const&, bool) ()`

and this line:

`reach::ReachStudy::ReachStudy(std::shared_ptr<reach::IKSolver const>, std::shared_ptr<reach::Evaluator const>, std::shared_ptr<reach::TargetPoseGenerator const>, std::shared_ptr<reach::Display const>, std::shared_ptr<reach::Logger>, reach::ReachStudy::Parameters) ()
`


However I am sure it is a scope problem coming from this line (as you described) because changing the reach code from

```

 reach::TargetPoseGenerator::ConstPtr target_pose_generator; 
 { 
   auto factory = loader.createInstance<TargetPoseGeneratorFactory>(get<std::string>(pose_gen_config, "name")); 
   target_pose_generator = factory->create(pose_gen_config); 
 } 
```
to:

```
auto factory = loader.createInstance<TargetPoseGeneratorFactory>(get<std::string>(pose_gen_config, "name")); 
 reach::TargetPoseGenerator::ConstPtr target_pose_generator; 
 { 
   target_pose_generator = factory->create(pose_gen_config); 
 } 
```

Fixes the problem. I have no clue why the pointer get out of scope since I am doing exactly the same thing than de `point_cloud_target_pose_generator`

**Reply by marip8** [reply]
That does seem odd. I would have to attach a debugger and step through the code to understand why it's not working as expected. Can you share the repo where you are building this? I have not had an issue building plugins outside the `reach` repo (e.g., `reach_ros`, `reach_ros2`, and some private repos I can't share), so I'm not sure what would be causing the issue

**Reply by vicltz** [reply]
[plugin.tar.gz](https://github.com/ros-industrial/reach/files/12418818/plugin.tar.gz)

I agree with you on the point that since `reach_ros` basically does that, this segfault should not happen.

Attached to this message you'll find the `cpp` and `h` file. I compile with the setup described above. If you can test it in another folder and tell me if you're able to reproduce the error it would be very nice.

And here you'll find a proper setup for the plugin : 

```
target_pose_generator:
  name: TestPlugin
  bounds_x: [-0,1]
  bounds_y: [-1,1]
  z: 0.5
  step: 0.1
  orientation: "60:30"
  shape: "plane"
```

Thank you for your help.
